tests/test_update.py

```diff
@@ -0,0 +1,91 @@
+ Secondary's package manager is set to `fake`
+ Primary goal is to verify whether aktualizr succeeds with a binary/fake update of secondary
+ while aktualizr/primary is configured with ostree package manager   
+"""
+@with_uptane_backend(start_generic_server=True)
+@with_director()
+@with_treehub()
+@with_sysroot()
+@with_secondary(start=True)
+@with_aktualizr(start=False, run_mode='once', output_logs=True)
+def test_primary_ostree_secondary_fake_updates(uptane_repo, secondary, aktualizr, director,
+                                               uptane_server, sysroot, treehub):
+    target_rev = treehub.revision
+    # add an ostree update for Primary
+    uptane_repo.add_ostree_target(aktualizr.id, target_rev)
+    # add a fake/binary update for Secondary
+    secondary_update_hash = uptane_repo.add_image(secondary.id, "secondary-update.bin")
+
+    with aktualizr:
+        aktualizr.wait_for_completion()
+
+    # check the Primary update, must be in pending state since it requires reboot
+    pending_rev = aktualizr.get_primary_pending_version()
+    if pending_rev != target_rev:
+        logger.error("Pending version {} != the target one {}".format(pending_rev, target_rev))
+        return False
+
+    # check the Secondary update
+    current_secondary_image_hash = aktualizr.get_current_image_info(secondary.id)
+    if current_secondary_image_hash != secondary_update_hash:
+        logger.error("Secondary current image {} != {} expected one".format(current_secondary_image_hash,
+                                                                            secondary_update_hash))
+        return False
+
+    # emulate reboot and run aktualizr once more
+    sysroot.update_revision(pending_rev)
+    aktualizr.emulate_reboot()
+
+    with aktualizr:
+        aktualizr.wait_for_completion()
+
+    # check the Primary update after reboot
+    result = director.get_install_result() and (target_rev == aktualizr.get_current_primary_image_info())
+    return result
+
+
+if __name__ == "__main__":
+    logging.basicConfig(level=logging.INFO)
+
+    parser = argparse.ArgumentParser(description='Test backend failure')
+    parser.add_argument('-b', '--build-dir', help='build directory', default='build')
+    parser.add_argument('-s', '--src-dir', help='source directory', default='.')
+
+    input_params = parser.parse_args()
+
+    KeyStore.base_dir = input_params.src_dir
+    initial_cwd = getcwd()
+    chdir(input_params.build_dir)
+
+    test_suite = [
+        test_primary_ostree_secondary_fake_updates
+    ]
+
+    test_suite_run_result = True
+    for test in test_suite:
+        logger.info('>>> Running {}...'.format(test.__name__))
+        test_run_result = test()
+        logger.info('>>> {}: {}'.format('OK' if test_run_result else 'Failed', test.__name__))
+        test_suite_run_result = test_suite_run_result and test_run_result
+
+    chdir(initial_cwd)
+    exit(0 if test_suite_run_result else 1)
+
```

Comment:
This is powerful and cool, but I wonder if it would've been possible to test this at a lower level. This is also becoming quickly redundant with `shared_cred_prov_test.py` and we should consider merging them at some point.

Also, minor nitpick: the name of this test/file is quite general. It's worth getting a bit more specific next time we mess with this.

Reply:
Yes, I agree, the name is too generic.

In terms of 'This is also becoming quickly redundant with shared_cred_prov_test.py',  I think they are very different, ' shared_cred_prov_test.py' tests just provisioning of non-ostree against the real backend, while 'test_update.py' tests a secondary update against the mocked backend if primary is based on ostree. This is the only test at the moment that can catch the issue described here https://saeljira.it.here.com/browse/OTA-3795, none of our tests could catch the issue before.

In retrospect, I think it makes sense to move this test to the IP secondary test suite (ipsecondary_test.py) as it actually tests the secondary update or replace it with lower level tests (C/C++ based) utilizing a virtual secondary. The later, will require duplication of the boilerplate code around ostree mocking located in `aktualizr_fullostree_test.cc` or the test logic itself can be added to aktualizr_fullostree_test.cc, what I tried but failed.

Reply:
> In terms of 'This is also becoming quickly redundant with shared_cred_prov_test.py', I think they are very different

Fair enough. Thanks for the explanation!

> I think it makes sense to move this test to the IP secondary test suite (ipsecondary_test.py) as it actually tests the secondary update

Not a bad idea.

> or replace it with lower level tests (C/C++ based) utilizing a virtual secondary

Yeah, we could do that, but I'd say that isn't a priority, since you already have a nice working version in Python!